Brave Today on New Tab Page

petemill opened the pull request:
Resolves https://github.com/brave/brave-browser/issues/12574

### TODO
- [x] UI for "update available" and fetching action
- [x] Publisher name style fix
- [x] Automatic feed update every 1 hour
- [x] Automatic sources update every 1 day
- [x] Categorize sources in "customize" panel
- [x] Deep-link to customize sources on new "customize" button whilst scrolling Brave Today
- [x] Fix overall NTP container
- [x] Default pref to show Brave Today only if browser is set to English language.
- [x] Lazy-load data and code on NTP
- [x] Squash
- [x] Clear feed cache when user performs history clear
- [x] Clear data on "Reset sources to default" button'
- [x] fix lint
- [x] strings
- [x] declare new deps
- [x] cr87 TrustedTypes fixes
- [x] production urls
- [x] UI for no data
- [x] UI for loading state

### Follow-up
- [ ] Finish publisher title context-menu for disabling source
- [ ] Consider not storing feed and publisher list in extension-process memory, only local storage. It's quite large.
- [ ] p3a questions
- [ ] japan feed
- [ ] sponsored items
- [ ] more unit tests
- [ ] don't read feed cache to memory if user has previously turned on brave today but now has it turned off

# Description

### Implementing Brave Today on Desktop:
- Fetching a single news feed from brave's "Private CDN" and displaying that feed on the NTP, mixing, choosing and paginating the feed accordingly. Using history data to score items based on whether the user has visited the article's Url host recently.
- Fetching padded images for the feed from Private CDN and decoding them, passing them to the front-end as a "data URI".
- Storing user preferences about which publishers (or sources) the user wishes to see feed items from and then performing that filtering on the device (in the browser).
- Displaying feed on the front-end on the New Tab Page. Scrolling down shows the UI. This is enabled by default for English language users and opt-in (via NTP settings) for everyone else.
- Adds Brave Today redux store including new action setup without constant strings _and_ new async redux middleware.

### Implementing a communication channel between WebUI (chrome://newtab) and the Brave Extension background page.

This is being used on this project for 2 reasons:

1. Because `SharedWorker` is not (easily) supported in WebUI and I did not want to perform feed and image data manipulation on the front-end JS thread and for every single NTP instance (which are frequent).

2. To perform a fetch of the feed, since NTP front-end is not able to make requests on the network.

Certainly, this could be performed in c++, but timing and the need to iterate meant that the background JS "worker" was a great place for this. This background fetching and prefs may well be ported to c++ in the future, most likely when working on bringing this feature to Android.

### Storage / prefs
Aside from the regular usage of chromium profile prefs to determine if the user has opted-out if Language is set to English or opted-in for all other Languages, we are using extension (local) storage for user preferences as to which publisher sources to enable or disable for the Brave Today feed.

## Architecture

Notable areas to consider:

- Extension background process
  - today/index.ts
    Sets up listening to messages from the NTP WebUI (via `chrome.runtime.onMessageExternal`)
  - today/feed.ts.
    Fetches and caches (both in-memory and in local storage) remote data from private CDN.
  - today/publishers.ts
    Fetches and caches (both in-memory and in local storage) remote publisher list from private CDN.
  - today/publisher-user-prefs.ts
    Sets and gets prefs about which publishers are explicitly enabled or disabled by the user. Stores in `chrome.storage.local`.
  - today/feedToData.ts
    Content selection algorithm. Breaks the articles in to categories and pages. Somewhat randomizes and adds weight based on the user's recent browsing history.
  - today/privateCDN.ts
   Expresses URLs to fetch content from, and helps "unpad" consistent-byte-length images.

- NTP WebUI renderer process
  - brave_new_tab_ui/async/today.ts
    Handles UI actions (init, refresh, preference-changed, etc) and communicates with extension background process, receiving feed data.
  - brave_new_tab_ui/components/default/braveToday/cards/CardImage.tsx
    Communicates with extension background process, receiving image data via padded content fetched from the Private CDN.


## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:


## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

petemill commented on components/brave_new_tab_ui/stories/default.tsx:
```diff
@@ -19,8 +19,27 @@ import * as binanceActions from '../actions/binance_actions'
+// TODO(petemill): privateCDN should be in /common/
+// import { getUnpaddedAsDataUrl } from '../../brave_extension/extension/brave_extension/background/today/privateCDN'
+
+// TODO(petemill): Have the private CDN contain a CORS response so we can directly fetch
```
Just to be clear, this is a storybook page meaning that this code is not bundled and delivered to users. It runs on dev machines and a static website, as a development and test tool. So the mentions of using this random 3rd party `cors-anywhere` site should be temporary and never for production code.

fmarier replied:
Fair enough. I also don't think that it's a problem for the private CDN to return CORS headers since everything hosted there is public and credential-less.

petemill commented on patches/extensions-common-api-_api_features.json.patch:
```diff
@@ -0,0 +1,12 @@
+     "contexts": ["webui"],
+     "matches": [
+       "chrome://password-change/*"
++      ,"chrome://newtab/*"
```
Enables `chrome.runtime.sendMessage` from the New Tab Page @brave/patch-reviewers

mkarolin replied:
If this is ok with the security reviewers, the patch LGTM.

diracdeltas replied:
sgtm as long as the receiver validates `sender.id` in `chrome.runtime.onMessage` (to make sure it's coming from the expected extension)

diracdeltas replied:
i see this allows any `chrome://` URL - https://github.com/brave/brave-core/pull/6192/files#diff-4263de6a462535357156da5fb6faf5e8c5adaf5404bde45194da89d5e93f19c4R33-R38. this should be ok as long as extensions can't register `chrome://` URLs. it'd be more restrictive to filter it to just ID if possible.

petemill replied:
Well, there is no `id` since the NTP isn't an extension, so if we want to further restrict then we can also check the hostname (`newtab` in this case). But extensions or apps absolutely should not be able to register or run any code on `chrome://`.

diracdeltas replied:
sgtm then

diracdeltas commented on components/brave_extension/extension/brave_extension/manifest.json:
```diff
@@ -54,8 +59,8 @@
-  "permissions": [ "contentSettings", "settingsPrivate", "management", "tabs", "storage", "webNavigation", "contextMenus", "cookies", "*://*/*", "chrome://favicon/*" ],
-  "content_security_policy": "default-src 'self'; font-src 'self' data:; script-src 'self'; style-src 'unsafe-inline'; img-src 'self' data: chrome://favicon/;",
+  "permissions": [ "contentSettings", "settingsPrivate", "management", "tabs", "history", "storage", "webNavigation", "contextMenus", "cookies", "*://*/*", "chrome://favicon/*", "https://pcdn.brave.software/*", "https://brave-today-cdn.brave.software/*" ],
+  "content_security_policy": "default-src 'self' https://pcdn.brave.software https://brave-today-cdn.brave.software; font-src 'self' data:; script-src 'self'; style-src 'unsafe-inline'; img-src 'self' data: chrome://favicon/ https://pcdn.brave.software; connect-src https://pcdn.brave.software https://brave-today-cdn.brave.software;",
```
are we going to use brave.software instead of brave.com for this service? if so we might need to do key pinning for it and also add it to the network audit whitelist if it's not there already

fmarier replied:
Actually I think `brave-today-cdn.brave.software` will also change to `brave-today-cdn.brave.com` before this is merged.

petemill replied:
Still waiting for this to be deployed. It's coming soon as far as I can tell.

fmarier replied:
Just to recap, once the files have been uploaded to the right places, we can:
- put `pcdn.brave.com` and `brave-today-cdn.brave.com` in `connect-src` and `img-src`
- leave `default-src` set to `'self'`

petemill replied:
@fmarier I've put the staging URLS in `connect-src` and removed any additions to `img-src` and `permissions`. PTAL.

bsclifton commented on browser/browsing_data/BUILD.gn:
```diff
@@ -12,6 +12,7 @@ source_set("browsing_data") {
   deps = [
     "//base",
+    "//brave/common/extensions/api",
     "//chrome/common",
```
I think you need to add a dep for `//extensions/browser` too - since `event_router.h` is included

petemill replied:
thanks - I also surrounded in a `if (enable_extensions) ...`

bsclifton commented on browser/ui/webui/brave_new_tab_message_handler.cc:
```diff
@@ -406,6 +412,8 @@ void BraveNewTabMessageHandler::HandleSaveNewTabPagePref(
     settingsKey = kNewTabPageShowClock;
   } else if (settingsKeyInput == "showStats") {
     settingsKey = kNewTabPageShowStats;
+  } else if (settingsKeyInput == "showToday") {
```
I know this is just adding; but it might be good to have a proper map here. We can do something like:
```cpp
#include "base/containers/flat_map.h"
#include "base/no_destructor.h"

//..

void BraveNewTabMessageHandler::HandleSaveNewTabPagePref(
    const base::ListValue* args) {

    //..

    static const base::NoDestructor<base::flat_map<const char*, const char*>> bool_mappings({
        {"showBackgroundImage", kNewTabPageShowBackgroundImage},
        {"brandedWallpaperOptIn", kNewTabPageShowSponsoredImagesBackgroundImage},
        //..
        {"showToday", kNewTabPageShowToday},
    });

    auto it = bool_mappings->find(settingsKeyInput);
    if (it == bool_mappings->end()) {
        LOG(ERROR) << "Invalid setting key: " << settingsKeyInput;
        return;
    }
    prefs->SetBoolean(it->second, settingsValueBool);
}
```

petemill replied:
I don't think I'll do that in this PR, there's too much going on here. Also I think there is an argument for removing this and using the `chrome.settingsPrivate` API.

bsclifton replied:
OK SGTM 👍

bsclifton commented on components/brave_extension/extension/brave_extension/background/today/feedToData.ts:
```diff
@@ -0,0 +1,260 @@
+import { formatDistanceToNow } from 'date-fns'
+import hashStrings from '../../../../../common/hashStrings'
+
+function shuffleArray(array: Array<any>) {
```
Is there any spec for the shuffle? ex: what type of shuffle algo is this (is that good enough?)

petemill replied:
nope, just random AFAIK. It's only 4 articles per page that are selected at random from the past 48 hours. That happens here:

```
const randomArticles = take(articles, 4, isArticleWithin48Hours, true)
```

bsclifton commented on components/brave_extension/extension/brave_extension/background/today/feedToData.ts:
```diff
@@ -0,0 +1,260 @@
+    }
+  }
+
+  // Ordered by # of occurances
```
s/occurances/occurrences

petemill replied:
fixed

bsclifton commented on components/brave_extension/extension/brave_extension/background/today/feedToData.ts:
```diff
@@ -0,0 +1,260 @@
+  }
+}
+
+export default async function getBraveTodayData (
```
Lots of logic in here - could be useful to break specific parts out to a new method(s) if we need to unit test specific parts. That could always be done at a later time (follow up) 😄

petemill replied:
Yeah I was thinking I'd unit test this whole function. Already broken out `take<T>`, `generateNextPage` and the weighting, but can also break out all the category calculations.

bsclifton commented on components/brave_extension/extension/brave_extension/background/today/feedToData.ts:
```diff
@@ -0,0 +1,260 @@
+    let score = article.score
+    const hasDomainHistoryMatch = historyHosts.includes(domainFromUrlString(article.url))
+    if (hasDomainHistoryMatch) {
+      score -= 5
```
What is the significance of 5? (what is a max score? etc)

petemill replied:
Spec determines 5. iOS has the same hardcoded value. Works with the initial score determined server-side.

bsclifton commented on components/brave_extension/extension/brave_extension/background/today/index.ts:
```diff
@@ -0,0 +1,172 @@
+// It handles command communication from front-ends, as well
+// as setting up and handling timers, and events from the
+// rest of the browser.
+//
```
Nice comments here (and in other files too)! 😻 Definitely helps when reading this

bsclifton commented on components/brave_extension/extension/brave_extension/background/today/publishers.ts:
```diff
@@ -0,0 +1,125 @@
+  await getLocalDataLock
+  if (memoryData) {
+    return memoryData
+  } else {
```
maybe a stylistic thing - so feel free to ignore 😛 But when doing a boolean check, it can sometimes be more readable to just do the second action instead of putting into an else. ex:
```
if (memoryData) {
    return memoryData
}
return await update()
```

petemill replied:
Thanks for spotting, probably just how it iterated in this case

bsclifton commented on components/brave_new_tab_ui/components/default/braveToday/index.tsx:
```diff
@@ -0,0 +1,87 @@
+    const options = { root: null, rootMargin: '0px', threshold: 0.25 }
+
+    this.braveTodayHitsViewportObserver = new
+      IntersectionObserver(this.handleBraveTodayHitsViewportObserver, options)
```
nice uses of intersection observers 😄 👍  I remember when those were added (Chromium 52? was before we hard forked Electron)

petemill replied:
This one is @cezaraugusto !

bsclifton commented on components/brave_new_tab_ui/containers/newTab/index.tsx:
```diff
@@ -502,6 +512,15 @@ class NewTabPage extends React.Component<Props, State> {
 
+  onBraveTodayInteracting = (isInteracting: boolean) => {
+    // TODO(petemill): Detect percentage of scroll towards brave today
+    // and set opacity at that level, instead of a binary on/off.
```
that would be cool! 😻 is it possible to do gradients with opacity?

petemill replied:
Ah this is already done, just not pushed. Not sure if I like how it feels actually, but will see.

bsclifton commented on components/definitions/today.d.ts:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  export interface Page {
+    articles: Article[] // 13
```
What are the numbers here? The maximum number of entries?

petemill replied:
It was just a helpful indicator to show how many are expected to be put in each of those arrays, since you can't easily specify array size in typescript (even if we wanted to). I can make it clearer.

petemill commented on package.json:
```diff
@@ -272,6 +274,7 @@
     "commander": "^2.9.0",
     "css-loader": "^2.1.1",
     "csstype": "^2.5.5",
+    "date-fns": "^2.15.0",
```
New deps declared. I did have to modify `check_npm_licenses.py` to support scoped repos. In general as a follow-up we need to move all deps from devDependencies to dependencies IMO.

petemill commented on components/brave_extension/extension/brave_extension/background/today/feed.ts:
```diff
@@ -0,0 +1,116 @@
+import { isPublisherContentAllowed } from '../../../../../common/braveToday'
```
fixed

petemill commented on components/brave_extension/extension/brave_extension/background/today/feedToData.ts:
```diff
@@ -0,0 +1,260 @@
+  const pages: BraveToday.Page[] = []
+  let canGenerateAnotherPage = true
+  // Sanity check: arbitrary max pages so we don't end up
+  // in infinit loop.
```
fixed

petemill commented on components/brave_extension/extension/brave_extension/background/today/privateCDN.ts:
```diff
@@ -0,0 +1,22 @@
+
+export async function getUnpaddedAsDataUrl(buffer: ArrayBuffer, mimeType = 'image/jpg') {
+  const data = new DataView(buffer)
+  const contentLength = data.getUint32(0, false)
```
I added that, thanks for the tip

petemill commented on components/brave_extension/extension/brave_extension/background/today/privateCDN.ts:
```diff
@@ -0,0 +1,22 @@
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+export const URLS = {
+  braveTodayFeed: 'https://pcdn.brave.software/brave-today/feed.json',
```
Still waiting on this to be pushed to *.brave.com

petemill commented:
OK this is ready for re-review. Merge blockers are:

- l10n strings
- production URLs

But if I can get some approvals (or re-reviews) then I can merge ASAP when those quick things are done.

fmarier commented on components/brave_new_tab_ui/actions/today_actions.ts:
```diff
@@ -0,0 +1,44 @@
+import { createAction } from 'redux-act'
```
Missing copyright header.

fmarier commented on components/brave_new_tab_ui/async/today.ts:
```diff
@@ -0,0 +1,87 @@
+import AsyncActionHandler from '../../common/AsyncActionHandler'
```
Copyright header

fmarier commented on components/brave_new_tab_ui/reducers/today/index.ts:
```diff
@@ -0,0 +1,99 @@
+import { createReducer } from 'redux-act'
```
Copyright header

fmarier commented on components/definitions/react-virtualized-auto-sizer.d.ts:
```diff
@@ -0,0 +1,38 @@
+declare module '@brave/react-virtualized-auto-sizer' {
```
Copyright header

fmarier commented on components/definitions/today.d.ts:
```diff
@@ -0,0 +1,108 @@
+declare namespace BraveToday {
```
Copyright header

fmarier commented on third_party/npm_redux-act/README.chromium:
```diff
@@ -0,0 +1,4 @@
+Name: redux-act
+URL: https://github.com/pauldijou/redux-act
+License: Apache License 2.0
```
I haven't found a good place to document this, but I've been trying to keep these in the standard [SPDX format](https://spdx.org/licenses/) in case we want to programmatically check licenses in the future. That one one would be `Apache-2.0`

petemill replied:
thanks, updated

fmarier reviewed (commented):
Looks good to me. I'd like to review the CSP changes once more before you merge, but everything else looks great. I only left a few minor comments.

bsclifton commented:
On Windows (and likely Linux), the scrollbars are a bit close to the toggles in the `Customize` modal:
![image](https://user-images.githubusercontent.com/4733304/98085917-438ab680-1e3b-11eb-8287-474f470ca8e0.png)
![image](https://user-images.githubusercontent.com/4733304/98086002-5f8e5800-1e3b-11eb-9469-6c063932ec6e.png)

petemill commented:
> On Windows (and likely Linux), the scrollbars are a bit close to the toggles in the Customize modal:

Yeah this is all platforms since @ryanml needed to have the Settings -> More Cards content go _underneath_ the scroll bar for some reason. We can fix hopefully in a follow-up.

petemill commented:
> Clicking Reset Brave Today settings shows a confirm, which has newtab says which is a bit awkward. I'm not sure we have control over editing this, though ☹️

We can file an issue to use a custom html dialog. Seems like a lot of effort for something minor, but can do.

bsclifton commented:
@petemill we can create issues - or if we're already planning to address, not a problem 😄 Both issues you responded too are not blockers. The customize button might have a bug though (for scenario described)

bsclifton reviewed (approved):
Looks really good! Left one comment (minor possible bug w/ customize). But everything else works great!

ryanml commented on components/brave_new_tab_ui/components/default/braveToday/cards/_articles/cardArticleLarge.tsx:
```diff
@@ -0,0 +1,89 @@
+const CardSingleArticleLarge = React.forwardRef<HTMLElement, Props>(function (props, ref) {
+  // no full content no render®
+  if (props.content.length === 0) {
+    return <></>
```
nit: any reason to not just return `null` here?

ryanml commented on components/brave_new_tab_ui/components/default/braveToday/cards/_articles/cardArticleLarge.tsx:
```diff
@@ -0,0 +1,89 @@
+      {props.content.map((item, index) => {
+        // If there is a missing item, return nothing
+        if (item === undefined) {
+          return <></>
```
same as above

ryanml commented on components/brave_new_tab_ui/components/default/braveToday/cards/cardIntro.tsx:
```diff
@@ -0,0 +1,28 @@
+import { getLocale } from '../../../../../common/locale'
+
+class HeaderBlock extends React.PureComponent<{}, {}> {
+  onClickHideWidget = () => {
```
is this being used?

petemill replied:
Huh... I didn't notice that. Not my code, but assume the plan was to be able to remove the intro card. Oh well, it's not hideable for this PR :-) Removed.

ryanml commented on components/brave_new_tab_ui/containers/newTab/index.tsx:
```diff
@@ -1161,28 +1178,57 @@ class NewTabPage extends React.Component<Props, State> {
           </Page.Footer>
+          {newTabData.showToday &&
+          <Page.GridItemNavigationBraveToday>
+            <p>Scroll for Brave Today</p>
```
i think this should be localized

petemill replied:
all localization done

ryanml commented on components/brave_new_tab_ui/containers/newTab/index.tsx:
```diff
@@ -1161,28 +1178,57 @@ class NewTabPage extends React.Component<Props, State> {
+          onAnotherPageNeeded={this.props.actions.today.anotherPageNeeded}
+          onInteracting={this.onBraveTodayInteracting}
+          // tslint:disable-next-line:jsx-no-lambda
+          onCustomizeBraveToday={() => { this.openSettings(SettingsTabType.BraveToday) }}
```
nit: why can't this be `onCustomizeBraveToday={this.openSettings.bind(this, SettingsTabType.BraveToday)}`? Then no ts-lint-ignore needed

petemill replied:
More readable. Both are the same performance impact (i.e. none). I assumed that if we had `jsx-no-lambda` we also would have `jsx-no-bind`. I think we should just disable `jsx-no-lambda` globally since if we're down to the creation of an anonymous function having a performance impact we're either doing something very right or _very_ wrong!

ryanml commented on components/brave_new_tab_ui/containers/newTab/settings/braveToday/sources.tsx:
```diff
@@ -0,0 +1,134 @@
+  // No publishers
+  // TODO(petemill): error state
+  if (!props.publishers) {
+    return <div>Loading...</div>
```
should be translatable

petemill commented:
OK all code and style feedback addressed. Strings are localized. Added Loading and Error states. Added Storybook pages for those individual UI states.

Will post detailed Brave Browser issue and test plan before merge. WIll update this PR with screenshots too.

petemill commented on browser/brave_profile_prefs.cc:
```diff
@@ -250,6 +255,15 @@ void RegisterProfilePrefs(user_prefs::PrefRegistrySyncable* registry) {
+  // english-language on browser startup.
+  const std::string locale =
+      brave_l10n::LocaleHelper::GetInstance()->GetLocale();
+  const std::string country_code = brave_l10n::GetLanguageCode(locale);
```
good suggestion, done

fmarier commented on components/brave_new_tab_ui/stories/default/data/todayStorybookState.ts:
```diff
@@ -0,0 +1,957 @@
+        category: `Top News`,
+        publish_time: `2020-10-22 00:09:00 UTC`,
+        url: `https://www.npr.org/2020/10/21/926445682/u-s-blames-iran-for-threatening-election-emails-says-russia-may-interfere-too?utm_medium=RSS&utm_campaign=news`,
+        img: `https://pcdn.brave.software/brave-today/cache/9bbfe86e499867a3988d7b089daa757cb7762a6a9d9f3b9836c4f5b8d8cd610b.jpg.pad`,
```
The `img` field is no longer included in the feed (https://github.com/brave-intl/brave-today-aggregator/pull/20).

fmarier commented on components/brave_new_tab_ui/stories/default/data/todayStorybookState.ts:
```diff
@@ -0,0 +1,957 @@
+        points: 18.525096955122685,
+        relative_time: `about 4 hours ago`
+      },
+      featuredDeals: [
```
Two small suggestions to make sure the client parsing code is robust:
- Include an article with `padded_img: ""` since that's a valid case when an image cannot be found.
- Include an article without the `padded_img` field at all, in case the backend stops emitting empty fields in the future.

fmarier reviewed (approved):
This is solid work @petemill !

petemill commented:
CI mostly passed
- ios ✅ 
- windows ✅ 
- linux ✅ 
- macos ⚠️ 🆗 
    - `test-install` failed 
- android ⚠️ 
    - random `npm init` error
        ```
        17:44:01  fatal: cannot lock ref 'refs/heads/saved-by-gclient-51af4604a': unable to resolve reference 'refs/heads/saved- 
        by-gclient-51af4604a': reference broken
        ```

petemill commented:
Pushed a very minor amend to fix a missing string and a bug that was preventing some deals from showing
https://github.com/brave/brave-core/compare/430287f635ef1517517c63df22c99dddc30ed7e5..01fe9f4e6d2129a35bc7295e23afbb51b9ef390d

...and also another rebase on master...

bsclifton commented:
There was [one browser-test failure on Linux](https://ci.brave.com/job/pr-brave-browser-brave-today-linux/22/execution/node/249/log/), but it appears to be intermittent [as it passed great on prior runs](https://ci.brave.com/job/pr-brave-browser-brave-today-linux/21/execution/node/249/log/):
```
22:40:40  1 test failed:
22:40:40      BraveShieldsAPIBrowserTest.AllowScriptsOnceIframe (../../brave/browser/extensions/api/brave_shields_api_browsertest.cc:185)
```

bridiver commented on components/common/AsyncActionHandler.ts:
```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2020 The Brave Authors. All rights reserved.
```
`common` has a specific meaning in chromium and this should not be at the top level of components like this
